### PR TITLE
One dimensional root fix

### PR DESCRIPTION
**Pull request by catompkins**
Split up previous pull:

-One-dimension root fixes in line with issue #108

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Can you rebase on master (and squash your `cargo fmt` commits) please?

**Comment by catompkins** [comment]
> Can you rebase on master (and squash your `cargo fmt` commits) please?

done

File: src/types/roots.rs [hunk 1]
```diff
@@ -263,76 +247,54 @@ impl RootFdfSolver {
...
-        unsafe extern "C" fn inner_f<F: Fn(f64) -> f64>(
-            x: c_double,
-            params: *mut c_void,
-        ) -> c_double {
-            let params: &(*const F, *const (), *const ()) =
-                &*(params as *const (*const F, *const (), *const ()));
-            let f = &*params.0;
+        // convert rust functions to C
+        unsafe extern "C" fn inner_f<'a, F: Fn(f64) -> f64 + 'a>(
+            x: f64,
```
Comment (GuillaumeGomez) [inline src/types/roots.rs:264]: Please keep the libc types.

File: src/types/roots.rs [hunk 2]
```diff
@@ -263,76 +247,54 @@ impl RootFdfSolver {
...
-        unsafe extern "C" fn inner_df<DF: Fn(f64) -> f64>(
-            x: c_double,
-            params: *mut c_void,
-        ) -> c_double {
-            let params: &(*const (), *const DF, *const ()) =
-                &*(params as *const (*const (), *const DF, *const ()));
-            let df = &*params.1;
+
+        unsafe extern "C" fn inner_df<'a, DF: Fn(f64) -> f64 + 'a>(
+            x: f64,
```
Comment (GuillaumeGomez) [inline src/types/roots.rs:272]: Please keep the libc types.

File: src/types/roots.rs [hunk 3]
```diff
@@ -263,76 +247,54 @@ impl RootFdfSolver {
...
+            params: *mut ::std::os::raw::c_void,
+        ) -> f64 {
+            let df: &DF = &*(params as *const DF);
             df(x)
         }
+
         unsafe extern "C" fn inner_fdf<FDF: Fn(f64, &mut f64, &mut f64)>(
-            x: c_double,
+            x: f64,
```
Comment (GuillaumeGomez) [inline src/types/roots.rs:280]: Please keep the libc types.

File: src/types/roots.rs [hunk 4]
```diff
@@ -263,76 +247,54 @@ impl RootFdfSolver {
...
     ) -> ::Value {
-        unsafe extern "C" fn inner_f<F: Fn(f64) -> f64>(
-            x: c_double,
-            params: *mut c_void,
-        ) -> c_double {
-            let params: &(*const F, *const (), *const ()) =
-                &*(params as *const (*const F, *const (), *const ()));
-            let f = &*params.0;
+        // convert rust functions to C
+        unsafe extern "C" fn inner_f<'a, F: Fn(f64) -> f64 + 'a>(
```
Comment (GuillaumeGomez) [inline src/types/roots.rs:263]: You need to store `f`, `df` and `fdf` here as well, otherwise they will be called outside of this scope.

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Thanks!

**Pull request merged**